Sources/DNSClient/DNSClient+Query.swift

```diff
@@ -108,29 +114,62 @@ extension DNSClient {
             let question = QuestionSection(labels: labels, type: type, questionClass: .internet)
             let message = Message(header: header, questions: [question], answers: [], authorities: [], additionalData: [])
             
-            return self.send(message)
+            return self.send(message, to: nil, timeout: timeout)
         }
     }
 
+    // MARK: - Transport primitive (timeout-aware overload + wrapper)
+    
+    /// Historical behavior wrapper (30s default); forwards to timeout-aware overload.
+    /// Keeping this avoids any source change for existing callers.
     func send(_ message: Message, to address: SocketAddress? = nil) -> EventLoopFuture<Message> {
+        return self.send(message, to: address, timeout: .seconds(30))
+    }
+    
+    /// Timeout-aware transport primitive with proper cancellation and cache cleanup.
+    ///
+    /// Timers are canceled on success, and in‑flight entries are removed on timeout.
+    ///
+    /// - Parameters:
+    ///   - message: The complete DNS `Message` to be sent as a query.
+    ///   - address: The destination `SocketAddress` for this specific query. If `nil`, the client's
+    ///              default server address is used.
+    ///   - timeout: The maximum `TimeAmount` to wait for a response before the returned future
+    ///              fails with a timeout error.
+    /// - Returns: An `EventLoopFuture<Message>` that will be fulfilled with the server's
+    ///           response message, or fail if an error occurs (e.g., a timeout).
+    func send(_ message: Message, to address: SocketAddress? = nil, timeout: TimeAmount) -> EventLoopFuture<Message> {
         let promise: EventLoopPromise<Message> = loop.makePromise()
         
         return loop.flatSubmit {
+            // Register in-flight
             self.dnsDecoder.messageCache.withLockedValue { cache in
                 cache[message.header.id] = SentQuery(message: message, promise: promise)
             }
+            
+            // Write on the channel
             self.channel.writeAndFlush(message, promise: nil)
             
             struct DNSTimeoutError: Error {}
             
-            self.loop.scheduleTask(in: .seconds(30)) {
+            // Schedule a timeout that also removes the in-flight cache entry to avoid leaks
+            let timeoutTask = self.loop.scheduleTask(in: timeout) { [weak self] in
+                self?.dnsDecoder.messageCache.withLockedValue { cache in
```

Comment:
Oh good spot! One small nit: Can we make this a strong capture over `dnsDecoder.messageCache`? I'm not very fond of weakly capturing values as it adds some complexity to the logic. However, we can capture specific values in Swift - so I don't think we realistically ever need `[weak self]` in Swift Server code.

```suggestion
            let timeoutTask = self.loop.scheduleTask(in: timeout) { [messageCache = dnsDecoder.messageCache] in
                messageCache.withLockedValue { cache in
```